File: day2/main.py
reports: list[list[int]] = []

with open("./puzzle_input.txt", "r") as file_input:
    line = file_input.readline().strip()
    while line is None or line != "":
        reports.append([int(i) for i in line.split(" ")])
        line = file_input.readline().strip()

# reports = [
#     [7, 6, 4, 2, 1],
#     [1, 2, 7, 8, 9],
#     [9, 7, 6, 2, 1],
#     [1, 3, 2, 4, 5],
#     [8, 6, 4, 4, 1],
#     [1, 3, 6, 7, 9],
# ]


def report_is_unidirectional(rep: list[int]):
    is_ascending = rep[0] < rep[1]

    is_unidirectional = True
    for i in range(0, len(rep) - 1):
        if is_ascending:
            if rep[i] > rep[i + 1]:
                is_unidirectional = False
                break

        else:
            if rep[i + 1] > rep[i]:
                is_unidirectional = False
                break

    return is_unidirectional


def steps_are_within_range(steps: list[int]):
    for s in steps:
        if s not in [1, 2, 3]:
            return False

    return True


def report_is_safe(report: list[int]):
    if not report_is_unidirectional(report):
        return False

    steps: list[int] = []
    for i in range(0, len(report) - 1):
        steps.append(abs(report[i] - report[i + 1]))
        if len(steps) == 0:
            continue

        if not steps_are_within_range(steps):
            return False

    return True


def try_report_correction(report: list[int]):
    for i in range(len(report)):
        corrected = report.copy()
        corrected.pop(i)

        if report_is_safe(corrected):
            return corrected

    return None


def get_safe_reports_amount(report_list: list[list[int]], with_dampener=False) -> int:
    safe = 0

    for report in report_list:
        if report_is_safe(report):
            safe += 1
        else:
            if not with_dampener:
                continue

            corrected_report = try_report_correction(report)
            if corrected_report is not None:
                safe += 1

    return safe


safe_reports = get_safe_reports_amount(reports.copy())
print("Safe reports:", safe_reports)

dampener_safe_reports = get_safe_reports_amount(reports.copy(), with_dampener=True)
print("(Dampened) Safe reports:", dampener_safe_reports)

File: day2/Main.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class Main {
	public static void main(String[] args) {
		File fileInput = new File("./puzzle_input.txt");
		List<List<Integer>> reports = parseInput(fileInput);

		int safeReports = getSafeReportsAmount(reports);
		int dampenedSafeReports = getSafeReportsAmount(reports, true);
		System.out.println("Safe reports: " + safeReports);
		System.out.println("(Dampened) Safe reports: " + dampenedSafeReports);
	}

	private static int getSafeReportsAmount(List<List<Integer>> reports) {
		return getSafeReportsAmount(reports, false);
	}

	private static int getSafeReportsAmount(List<List<Integer>> reports, boolean withDampener) {
		int safeReports = 0;

		for (List<Integer> report : reports) {
			if (reportIsSafe(report)) {
				safeReports++;
			} else {
				if (!withDampener)
					continue;

				Optional<List<Integer>> correctedReport = tryReportCorrection(report);
				if (correctedReport.isPresent())
					safeReports++;
			}
		}

		return safeReports;
	}

	private static Optional<List<Integer>> tryReportCorrection(List<Integer> report) {
		for (int i = 0; i < report.size(); i++) {
			List<Integer> corrected = new ArrayList<>(report);
			corrected.remove(i);

			if (reportIsSafe(corrected))
				return Optional.of(corrected);
		}

		return Optional.empty();
	}

	private static boolean reportIsSafe(List<Integer> report) {
		if (!reportIsUnidirectional(report))
			return false;

		List<Integer> steps = new ArrayList<>();
		for (int i = 0; i < report.size() - 1; i++) {
			steps.add(Math.abs(report.get(i) - report.get(i + 1)));
			if (i == 0)
				continue;

			if (!isWithinRange(steps))
				return false;
		}

		return true;
	}

	private static boolean isWithinRange(List<Integer> steps) {
		final int[] allowedSteps = new int[] { 1, 2, 3 };
		return steps.stream().allMatch(s -> Arrays.stream(allowedSteps).anyMatch(v -> v == s));
	}

	private static boolean reportIsUnidirectional(List<Integer> report) {
		boolean isAscending = report.get(0) < report.get(1);

		for (int i = 0; i < report.size() - 1; i++) {
			if (isAscending) {
				if (report.get(i) > report.get(i + 1))
					return false;
			} else {
				if (report.get(i + 1) > report.get(i))
					return false;
			}
		}

		return true;
	}

	static List<List<Integer>> parseInput(File input) {
		try (BufferedReader reader = new BufferedReader(new FileReader(input))) {
			List<List<Integer>> reports = new ArrayList<>();
			while (true) {
				String line = reader.readLine();
				if (line == null)
					break;
				reports.add(parseReportLevels(line));
			}

			return reports;
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	private static List<Integer> parseReportLevels(String line) {
		String[] levels = line.split(" ");
		return Arrays.stream(levels).map(Integer::parseInt).collect(Collectors.toList());
	}
}